### ISSUE TITLE
Handling Per-Epoch Training with Grain Dataloader

### ISSUE THREAD

**Issue by danbnyn** [google/grain#572]
Hi there,

I’m currently working with the Grain Dataloader and encountered an issue with stopping and resuming training at each epoch boundary when initializing the sampler with `num_epochs > 0`. I’d like some guidance on how to handle this scenario properly and understand the shuffling mechanism across epochs.

### The Issue:
I’m using the `IndexSampler` with `num_epochs > 0`, and my expectation is to stop the dataloader at the end of each epoch and resume training at the next epoch. However, the sampler seems to provide a total number of items equal to `num_samples * num_epochs`, making it behave as if all epochs are combined into a single stream.

Here’s how I understand the current behavior:
- The `IndexSampler` creates a `MapDataset` with the number of records (`num_records`). This acts like an infinite list of record keys stitched together.
- When calling `__getitem__` with an index greater than `num_records` (but less than `num_epochs * num_records`), it acts like a modulus operation. This mechanism helps maintain access to one record per epoch.
- Shuffling appears to happen per slice of `num_records`, allowing random access across epochs while maintaining the structure.

### My Goal:
I want to stop at the end of the first epoch and then resume at the next epoch. My current implementation tracks the iterator state using `get_state()` and `set_state()`, but this feels somewhat clunky.

### What I’ve Tried:
1. **Tracking the iterator state:**
   - After each batch is processed, I use `get_state()` to check the `last_seen_indices`.
   
2. **Checking for the epoch boundary:**
   - I attempt to detect when the indices reach the end of the current epoch by comparing the `last_seen_indices` to a calculated boundary value based on `(current_epoch + 1) * num_records`.
   
3. **State update upon crossing the boundary:**
   - Once the boundary is crossed, I update the state by setting the indices 'last_seen_indices' just before the start of the next epoch for each worker and resetting `last_worker_index` to ensure workers resume correctly.

4. **Training loop logic:**
   - After each batch, I check the epoch boundary. If it is crossed, I break the loop, save the state, and later resume from this state.

However, this approach feels a bit manual, and I’m wondering if there’s a more optimal or intended way to handle this scenario, especially in terms of the shuffling behavior across epochs and stopping/resuming the dataloader.

### Questions:
1. **Is there a cleaner way to stop the dataloader at the end of each epoch and resume from the next?**
2. **Is my approach of tracking the iterator state with `get_state()` and `set_state()` appropriate for this use case, or is there a better approach?**

Thank you for your assistance!

**Comment by iindyk** [comment]
could you clarify what do you mean by 'dataloader stopping and resuming training at each epoch boundary'? What would you like the iterator to behave like? would you like it to raise a `StopIteration` at the epoch end? DataLoader shuffles every epoch differently and doesn't mix epochs. To replicate that and raise `StopIteration` you could create multiple loaders without repeat and pass different seeds to them. Or you could manually inspect the state like you do, but the state structure itself is not a public API, so it may change at some point.

**Comment by danbnyn** [comment]
I stumbled upon this issue while training a model as I wanted to evaluate it on each epoch and having to figure out a way to know when the epoch was finished (which is non trivial when you have multi workers) . A `StopIteration` would indeed fix my issue! 
About creating multiple Dataloader for each epoch, it feels wrong not to use the intended multi-epoch handling. On that topic maybe I'm doing something wrong but the initialisation of each Dataloader takes quite a while hence why I resorted to inspecting the state.

**Comment by iindyk** [comment]
The initialization results in restarting the worker pool and fetching the source metadata so it does have an overhead. `DataLoader` may be a bit too restrictive of an API for this. Try out our `{Map|Iter}Dataset` API, see [tutorial](https://github.com/google/grain/blob/main/tutorials/dataset_basic_tutorial.ipynb). With that API you will be able to use [map_with_index](https://github.com/google/grain/blob/5005e6b76adb1476ab0d9adf99dbf51fdb41d443/grain/_src/python/dataset/dataset.py#L396) from which you can calculate the epoch like this:

```

source = ...
num_elements = len(source)

def add_epoch_and_element_index(idx:int, element: dict):
  return {**element, "epoch": idx//num_elements, "index": idx%num_elements}

ds = grain.MapDataset.source(source)
ds = ds.map_with_index(add_epoch_and_element_index)
ds = ds.seed(42)
ds = ds.shuffle()
...
```

Then you can examine the produced element and the first element with change in "epoch" will mean the start of new epoch. The Dataset API is more low-level but allows a lot more flexibility. Make sure to use `mp_prefetch` in the end of the pipeline to enable multiprocess prefetching that you use with num_workers>0.

**Comment by danbnyn** [comment]
Thanks for providing those resources I'll test it out!
